File: java/strain/src/main/java/Strain.java
import java.util.LinkedList;
import java.util.List;
import java.util.function.Predicate;

public class Strain {
    public static <T> List<T> keep(List<T> input, Predicate<T> predicate) {
        LinkedList<T> result = new LinkedList<T>();

        for (T item: input) {
            if (predicate.test(item)) {
                result.add(item);
            }
        }

        return result;
    }

    public static <T> List<T> discard(List<T> input, Predicate<T> predicate) {
        return keep(input, predicate.negate());
    }
}

File: r/strain/strain.R
keep <- function(input, fun) {
  if (is.null(input)) return(c())
  
  kept <- vector(mode = mode(input))
  
  for (element in input) {
    if (fun(element)) {
      kept[[length(kept) + 1]] <- element
    }
  }
  
  kept
}

discard <- function(input, fun) {
  keep(input, Negate(fun))
}
